docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,479 @@
+
+**Note**: In order to complete the user story for all the providers, some of the advanced features (such as Extensibility/Transformation) are required. However, getting this in place even only for a subset of providers allows us to build and test a big chunk of the entire machinery.
+
+#### Story 3 - Easier UX for scaling workers nodes
+As an end user, I want to be able to easily scale up/down the number of replicas for each set of worker nodes in the cluster.
+- Currently, (for a cluster with 3 machine deployments) this is possible by updating these three different objects representing the sets of worker nodes in the pool. An easier user experience would be to update a single object to enable the scaling of multiple sets of worker nodes.
+
+### Implementation Details/Notes/Constraints
+
+The following section provides details about the introduction of new types and modifications to existing types to implement the ClusterClass functionality.
+If instead you are eager to see an example of ClusterClass and how the Cluster object will look like, you can jump to the Behavior paragraph.
+
+#### New API types
+##### ClusterClass
+This CRD is a collection of templates that describe the topology for one or more clusters.
+```golang
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed topologies.
+type ClusterClass struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+  // ControlPlane is a reference to a local struct that holds the details
+  // for provisioning the Control Plane for the Cluster.
+  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+  // Worker describes the worker nodes for the cluster.
+  // It is a collection of node types which can be used to create
+  // the worker nodes of the cluster.
+  // +optional
+  Worker WorkerClass `json:"worker,omitempty"`
+
+  // Infrastructure is a reference to a provider-specific template that holds
+  // the details for provisioning infrastructure specific cluster
+  // for the underlying provider.
+  // The underlying provider is responsible for the implementation
+  // of the template to an infrastructure cluster.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes which pronounce
+//
+type WorkerClass struct {
+  // Deployments is a list of machine deployment types that can be used to create
+  // a set of worker plane nodes.
+  Deployments []DeploymentClass `json:"deployments,omitempty"`
```

Comment:
I'd rather avoid using the term deployment/deploymentClass since there's already a kubernetes resource with that same name.

Reply:
MachineDeploymentClass??
But we can eventually reiterate during implementation and update the proposal accordingly